many minor edits and revision suggestions (#23)
bluegenes/2020-gep: content/04.workflows-and-software.md

If your use case extends beyond these tools, there are several scriptable workflow systems that offer comparable benefits for data intensive biology.
Each has it own strengths, meaning each software will meet an individuals computing goals differently (see **Table @tbl:workflows**).
Our lab has adopted Snakemake, in part due to its similarity and integration with Python, its flexibility for building and testing new analyses in different languages, and its intuitive integration with software management tools (described below)[@doi:10.1093/bioinformatics/bts480].

> Software like Nextflow and Common Workflow Language require slightly more infrastructure, but in return, scale much better to pipelines with hundreds of thousands of steps and support containerization more rigidly (CTB what are you trying to say here??), making them ideal for production-level pipelines [@doi:10.1038/nbt.3820; @doi:10.6084/m9.figshare.3115156.v2].

Reviewer: ```suggestion
Snakemake and Nextflow are commonly used for designing new research pipelines, where flexibility and iterative, branching development is a key feature [@doi:10.1038/nbt.3820]. Common Workflow Language (CWL) and Workflow Description Language (WDL) are interchangeable formats that are more geared towards scalability, making them ideal for production-level pipelines with hundreds of thousands of samples [@doi:10.6084/m9.figshare.3115156.v2]. CWL and WDL are currently best built and used through wrapper tools such as Rabix or platforms such as Terra [@https://rabix.io; @https://terra.bio].
```

Just trying to give comparative merits between workflow systems. But now that I'm thinking about it, the "more infrastructure" part may only apply to nextflow pipelines considered for nf-core. Is there a better way to get across the difference that snakemake builds the entire dag first, which means that scaling up can currently be frustrating?

Author: I was specifically flagging the "support containerization more rigidly" text.

Author: I don't know what that phrase means :)

Author: I would ignore scaling question. Sorta off topic for this paper.

Reviewer 2: @olgabot, do you have any insight here? We want to make sure we highlight Nextflow and CWL, but other than scaling, we're not sure how to best present the differences or trade offs. I think it's true that NF and CWL are more common in production level code, but I don't think that's always true.

Author: Yes, I don't think that nextflow is particularly more scalable than snakemake. CWL and WDL runners tend to think more on the "let's run 100,000 jobs now" side and so they ARE (or should be) more scalable, but I don't know of citations or objective data.

Author: Maybe the way to proceed is to distinguish Nextflow and Snakemake as research workflow systems that we explicitly do recommend for research bioinformaticians doing bespoke research pipeline development, vs CWL and WDL which are interchange formats that I think are best built and used through tools (like Rabix/SBG, or Terra). We could run something by luiz, camille, or michael crusoe if we wanted to double-check the language.

Reviewer: gave it a shot! I'll make a note to specifically ask for language check/feedback on this section when we send out to coauthors

Author: nice!

### Wrangling Scientific Software

Analysis workflows commonly rely on multiple software packages to generate final results.
These tools are heterogeneous in nature: they are written by researchers working in different coding languages, with varied approaches to software design and optimization, and often for specific analysis goals.
Each program has a number of other programs it depends upon to function ("dependencies"), and as software changes over time to meet research needs, the results may change, even when run with identical parameters.
As a result, it is critical to take an organized approach to installing, managing, and keeping track of software and software versions.

> To meet this need, most workflow managers integrate with software management systems like conda, singularity, and docker [@doi:10.1038/s41592-018-0046-7; @doi:10.1371/journal.pone.0177459; @doi:10.5555/2600239.2600241]. (CTB: one of these citations fails.)

Reviewer: it's the docker citation (though it seems to be real) https://dl.acm.org/doi/10.5555/2600239.2600241

> To meet this need, most workflow managers integrate with software management systems like conda, singularity, and docker [@doi:10.1038/s41592-018-0046-7; @doi:10.1371/journal.pone.0177459; @doi:10.5555/2600239.2600241]. (CTB: one of these citations fails.)

Reviewer: ```suggestion
To meet this need, most workflow managers integrate with software management systems like conda, singularity, and docker [@doi:10.1038/s41592-018-0046-7; @doi:10.1371/journal.pone.0177459; @https://dl.acm.org/doi/10.5555/2600239.2600241]. (CTB: one of these citations fails.)
```

program, you can ensure all programs will run without issue. Using small, separate environments for your software and building many simple environments to accommodate different steps in your workflow also reduces the amount of time it takes conda to resolve dependency conflicts between different software tools ("solve" an environment). Conda virtual environments can be created and installed either on the command line, or via an environment YAML file, as shown. In this case, the environment file
also specifies which Conda channels to search and download programs from. When specified in a YAML file, conda environments are easily transferable between computers and operating systems. Further, because the version of each package installed in an environment is recorded, workflow reproducibility is enhanced. Although portions of conda may be superseded by alternative solutions [@url:https://github.com/QuantStack/mamba], this model of software installation and management will likely
persist.](images/conda_figure.svg){#fig:conda_figure height=8in}

#### Getting started with software management

> (CTB: the formatting of the heading above matches the formatting of the paragraph leads below. Hard to distinguish.)

Reviewer: ```suggestion
```

> **Using software without learning management systems**

Reviewer 2: ```suggestion
*Using software without learning management systems*
```
